### ISSUE TITLE
Random Unity crashes with "phonon.dll caused an Access Violation" after upgrade to 4.0.0

### ISSUE THREAD

**Issue by Frooxius** [ValveSoftware/steam-audio#157]
Hello!

First, congrats on the 4.0.0 version release, it's looking (well sounding) awesome! (also love the reason for the new version number).

We have been using Steam Audio with Unity probably over a year in our project with no issues. After I've upgraded our project to 4.0.0 version, we started getting reports of random crashes from some of our players, caused by Access Violation in phonon.dll

Below I've attached a few of the crash dumps that we collected from our users:
[Steam Audio 4.0.0 Crashes.zip](https://github.com/ValveSoftware/steam-audio/files/6911095/Steam.Audio.4.0.0.Crashes.zip)

Unfortunately I am not certain how to replicate this, as the crashes are seemingly random and only affect some users from what I can tell. I hope the crash dumps are sufficient, but let me know if there's anything more from my end that I could provide.

After downgrading back to 2.0-beta20 the crashes have stopped.

Most of the crashes seem to occur in the rtcUpdateBuffer function. The stacktraces look typically like this:

```
Stack Trace of Crashed Thread 8816:
0x00007FFD5FD79745 (phonon) rtcUpdateBuffer
0x00007FFD5FDE970E (phonon) embree::TaskScheduler::wait
0x00007FFD5FDE9C58 (phonon) embree::TaskScheduler::wait
0x00007FFD5FE1311F (phonon) iplContextRetain
0x00007FFDB68FC774 (audioplugin_phonon) UnityGetAudioEffectDefinitions
0x00007FFD893C2666 (UnityPlayer) UnityMain
0x00007FFD898C5CC8 (UnityPlayer) UnityMain
0x00007FFD898C596A (UnityPlayer) UnityMain
0x00007FFD898C596A (UnityPlayer) UnityMain
0x00007FFD898C596A (UnityPlayer) UnityMain
0x00007FFD898C596A (UnityPlayer) UnityMain
0x00007FFD898C596A (UnityPlayer) UnityMain
0x00007FFD898C596A (UnityPlayer) UnityMain
0x00007FFD898C596A (UnityPlayer) UnityMain
0x00007FFD898C596A (UnityPlayer) UnityMain
0x00007FFD898C596A (UnityPlayer) UnityMain
0x00007FFD898C596A (UnityPlayer) UnityMain
0x00007FFD898C596A (UnityPlayer) UnityMain
0x00007FFD898C596A (UnityPlayer) UnityMain
0x00007FFD899370E1 (UnityPlayer) UnityMain
0x00007FFD898E6269 (UnityPlayer) UnityMain
0x00007FFD89932394 (UnityPlayer) UnityMain
0x00007FFD898BE926 (UnityPlayer) UnityMain
0x00007FFD89D629E0 (UnityPlayer) UnityMain
0x00007FFDD5B27034 (KERNEL32) BaseThreadInitThunk
0x00007FFDD7762651 (ntdll) RtlUserThreadStart
```

**Comment by laurelhach** [comment]
Hi, 

I do have a lot of crash too when playing with the new version.
Ususally it happens when loading / unloading a level async.

`Stack Trace of Crashed Thread 40580:
0x00007FF80935D013 (audioplugin_phonon) UnityGetAudioEffectDefinitions
0x00007FF80935CEBC (audioplugin_phonon) UnityGetAudioEffectDefinitions
0x00007FFFBC40FF71 (UnityPlayer) UnityMain
0x00007FFFBC9EABFE (UnityPlayer) UnityMain
0x00007FFFBC427FF5 (UnityPlayer) UnityMain
0x00007FFFBC452ED0 (UnityPlayer) UnityMain
0x000001D82C7B8754 (UnityEngine.AudioModule) UnityEngine.AudioSource.SetSpatializerFloat()
0x000001D82C7B89E3 (SteamAudioUnity) SteamAudio.UnityAudioEngineSource.SetSpatializerIntPtr()
0x000001D82C7B857B (SteamAudioUnity) SteamAudio.UnityAudioEngineSource.UpdateParameters()
0x000001D834AD8468 (SteamAudioUnity) SteamAudio.SteamAudioSource.Update()
0x000001D72A09BB00 (mscorlib) System.Object.runtime_invoke_void__this__()
0x00007FFFC451E650 (mono-2.0-bdwgc) mono_get_runtime_build_info
0x00007FFFC44A2AE2 (mono-2.0-bdwgc) mono_perfcounters_init
0x00007FFFC44ABB3F (mono-2.0-bdwgc) mono_runtime_invoke
0x00007FFFBC2493FD (UnityPlayer) UnityMain
0x00007FFFBC246662 (UnityPlayer) UnityMain
0x00007FFFBC22E303 (UnityPlayer) UnityMain
0x00007FFFBC22E3DF (UnityPlayer) UnityMain
0x00007FFFBBFA3DF0 (UnityPlayer) UnityMain
0x00007FFFBC0D888A (UnityPlayer) UnityMain
0x00007FFFBC0D8930 (UnityPlayer) UnityMain
0x00007FFFBC0DC688 (UnityPlayer) UnityMain
ERROR: SymGetSymFromAddr64, GetLastError: 'Tentative d’accès à une adresse non valide.' (Address: 00007FFFBBE9EEAA)
0x00007FFFBBE9EEAA (UnityPlayer) (function-name not available)
ERROR: SymGetSymFromAddr64, GetLastError: 'Tentative d’accès à une adresse non valide.' (Address: 00007FFFBBE9D51D)
0x00007FFFBBE9D51D (UnityPlayer) (function-name not available)
ERROR: SymGetSymFromAddr64, GetLastError: 'Tentative d’accès à une adresse non valide.' (Address: 00007FFFBBEA2622)
0x00007FFFBBEA2622 (UnityPlayer) (function-name not available)
0x00007FFFBBEA35AB (UnityPlayer) UnityMain
ERROR: SymGetSymFromAddr64, GetLastError: 'Tentative d’accès à une adresse non valide.' (Address: 00007FF7AA6E11F2)
0x00007FF7AA6E11F2 (Ruzar 2019) (function-name not available)
0x00007FF8221D7034 (KERNEL32) BaseThreadInitThunk
0x00007FF8229C2651 (ntdll) RtlUserThreadStart`

Anyone can help?
I never had any issues with the beta version.

**Comment by lakulish** [comment]
@Frooxius Based on the crash dumps, it looks like Unity is asking Steam Audio to render binaural audio (which requires 2 channels) into a mono (1-channel) buffer. Is it possible for users to configure their output audio format to be mono?

@laurelhach Your crashes may be unrelated. Can you provide crash dumps and log files from Unity? Thanks!

**Comment by laurelhach** [comment]
@lakulish here is the rar file:
It contains a bunch of crashes

[laurel_SteamAudio4.0_Crashes.zip](https://github.com/ValveSoftware/steam-audio/files/6977865/laurel_SteamAudio4.0_Crashes.zip)

Thanks for your help!

**Comment by lakulish** [comment]
@laurelhach Thanks for providing the crash dumps. Most of them indicate that the crash occurs when Steam Audio tries to load scene data, so it's possible that async loads cause invalid data to be sent to Steam Audio somehow. 

However, I'm unable to reproduce this behavior with small test scenes that async load each other. Would it be possible for you to provide a test project or scene in which this issue can be somewhat reliably reproduced? If not, can you provide more details on how and when your project async loads a scene, are the scenes loaded additively, etc.? Also, how frequently do the crashes occur? Almost every time that an async load happens, or only rarely?

**Comment by laurelhach** [comment]
@lakulish I'll try to create a scene but I'm not sure I'll be able to get something (as my project is quite complexe).

In  the meantime:
1. I load / unload scenes using the additive mode in low priority:
2. It doesn't crash everytime I load / unload the levels. I would say 1-2 out of 10. 
3. Sometimes it feels that it's not crashing right after streaming a level though (and more randomly).
4. Setting the Serializer plugin and the Ambisonic Decoder Plugin to None (in the Project settings) doesn't crash the editor / game anymore (even if the Steam Audio Source Awake Method is triggered).

Just a side note, I have 2 Audio listener in my project and by default, none are loaded. So when the main menu (additive) is loaded the audio listener is added (as I can see in your code, when a scene is loaded, the audio listener is re-assigned).

Here is your modified code (obivously, we should always have an audio listener, but if we don't, there is no check for null compoment in the SteamAudioManager.cs):
`
  void OnSceneLoaded(UnityEngine.SceneManagement.Scene scene, LoadSceneMode loadSceneMode)
  {
      LoadScene(scene, mContext, additive: (loadSceneMode == LoadSceneMode.Additive));

      mListener = AudioEngineStateHelpers.Create(SteamAudioSettings.Singleton.audioEngine).GetListenerTransform();

      if (mListener != null)
          mListenerComponent = mListener.GetComponent<SteamAudioListener>();
      else
          Debug.LogWarning("no mListenerComponent found!");
  }
`

My Streaming Volume Section:

`
[SerializeField]
private ThreadPriority threadPriorityLoadAsync = ThreadPriority.Low;

private IEnumerator LoadAsync(LevelManager.AllLevels level)
{
      Application.backgroundLoadingPriority = threadPriorityLoadAsync;

      AsyncOperation asyncOperation = SceneManager.LoadSceneAsync(level.ToString(), LoadSceneMode.Additive);
      asyncOperation.allowSceneActivation = false;

      while (asyncOperation.progress < 0.9f)
          yield return null;
      
      while (!PlayerMovement.instance.CanStreamLevel)
          yield return null;

      asyncOperation.allowSceneActivation = true;
}
`

**Comment by Frooxius** [comment]
@lakulish We do not have anything that would allow users to switch the output to Mono, it's set to Stereo in the Project Settings as the "Default Speaker Mode".

I'm not sure if there are some external ways to modify it though or if Unity auto-switches to it based on system settings. I've asked our users to check if their system has Mono audio output.

Also apologies for the late response! I'll let you know if I get a response from our users.

**Comment by lakulish** [comment]
We've just released Steam Audio 4.1.0, which includes various crash fixes: https://github.com/ValveSoftware/steam-audio/releases/tag/v4.1.0.

Try it out and let us know if these crashes persist in the new version.

**Comment by kisak-valve** [comment]
Closing pending feedback.
